File: libs/moqtail-ts/src/model/common/pair.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from './byte_buffer'
import {
  NotEnoughBytesError,
  LengthExceedsMaxError,
  KeyValueFormattingError,
  ProtocolViolationError,
} from '../error/error'

const MAX_VALUE_LENGTH = 2 ** 16 - 1 // 65535
const MAX_U64 = 2n ** 64n - 1n

/**
 * @public
 * Represents a key-value pair for MOQT protocol parameters.
 *
 * - If `typeValue` is **even**, the value is a varint (`bigint`).
 * - If `typeValue` is **odd**, the value is a binary blob (`Uint8Array`) with a maximum length of 65535 bytes.
 *
 * Use {@link KeyValuePair.tryNewVarInt} for varint pairs and {@link KeyValuePair.tryNewBytes} for blob pairs.
 */
export class KeyValuePair {
  /**
   * The key/type identifier for this pair.
   * - Even: value is a varint.
   * - Odd: value is a blob.
   */
  public readonly typeValue: bigint

  /**
   * The value for this pair.
   * - If `typeValue` is even: a varint (`bigint`).
   * - If `typeValue` is odd: a binary blob (`Uint8Array`).
   */
  public readonly value: bigint | Uint8Array

  /**
   * Constructs a new KeyValuePair.
   * @param typeValue - The key/type identifier.
   * @param value - The value (varint or blob).
   * @internal Use static factory methods instead.
   */
  private constructor(typeValue: bigint, value: bigint | Uint8Array) {
    this.typeValue = typeValue
    this.value = value
  }

  /**
   * Creates a new varint KeyValuePair.
   * @param typeValue - Must be even.
   * @param value - The varint value.
   * @returns A KeyValuePair with varint value.
   * @throws KeyValueFormattingError if typeValue is not even.
   */
  static tryNewVarInt(typeValue: bigint | number, value: bigint | number): KeyValuePair {
    const tv = typeof typeValue === 'number' ? BigInt(typeValue) : typeValue
    if (tv % 2n !== 0n) {
      throw new KeyValueFormattingError('KeyValuePair.tryNewVarInt')
    }
    const v = typeof value === 'number' ? BigInt(value) : value
    return new KeyValuePair(tv, v)
  }

  /**
   * Creates a new blob KeyValuePair.
   * @param typeValue - Must be odd.
   * @param value - The binary blob value.
   * @returns A KeyValuePair with blob value.
   * @throws KeyValueFormattingError if typeValue is not odd.
   * @throws LengthExceedsMaxError if value length exceeds 65535 bytes.
   */
  static tryNewBytes(typeValue: bigint | number, value: Uint8Array): KeyValuePair {
    const tv = typeof typeValue === 'number' ? BigInt(typeValue) : typeValue
    if (tv % 2n === 0n) {
      throw new KeyValueFormattingError('KeyValuePair.tryNewBytes')
    }
    const len = value.length
    if (len > MAX_VALUE_LENGTH) {
      throw new LengthExceedsMaxError('KeyValuePair.tryNewBytes', MAX_VALUE_LENGTH, len)
    }
    return new KeyValuePair(tv, value)
  }

  /**
   * Serializes this key-value pair to a frozen byte buffer.
   * Equivalent to {@link KeyValuePair.serializeDelta} with `prevType = 0n`.
   * @returns The serialized buffer.
   */
  serialize(): FrozenByteBuffer {
    return this.serializeDelta(0n)
  }

  /**
   * Serializes this pair's wire form, encoding the Type as a delta from
   * `prevType` (the type of the previous KVP in the same list, or 0n if this
   * is the first entry).
   * @param prevType - The type of the previous KVP in the same list.
   * @throws ProtocolViolationError if this pair's type is less than prevType
   *   (Delta Type is an unsigned varint, so the list must be sorted ascending).
   */
  serializeDelta(prevType: bigint): FrozenByteBuffer {
    if (this.typeValue < prevType) {
      throw new ProtocolViolationError(
        'KeyValuePair.serializeDelta',
        `type ${this.typeValue} is less than previous type ${prevType}; a KVP list must be sorted by ascending type to be delta-encoded`,
      )
    }
    const buf = new ByteBuffer()
    buf.putVI(this.typeValue - prevType)
    if (isVarInt(this)) {
      buf.putVI(this.value)
    } else if (isBytes(this)) {
      buf.putLengthPrefixedBytes(this.value)
    }
    return buf.freeze()
  }

  /**
   * Deserializes a KeyValuePair from a buffer.
   * Equivalent to {@link KeyValuePair.deserializeDelta} with `prevType = 0n`.
   * @param buf - The buffer to read from.
   * @returns The deserialized KeyValuePair.
   * @throws LengthExceedsMaxError if blob length exceeds 65535 bytes.
   * @throws NotEnoughBytesError if buffer does not contain enough bytes.
   */
  static deserialize(buf: BaseByteBuffer): KeyValuePair {
    return KeyValuePair.deserializeDelta(buf, 0n)
  }

  /**
   * Deserializes a wire-form KVP whose Type field is a delta from `prevType`
   * (the type of the previous KVP in the same list, or 0n if this is the
   * first entry).
   * @param buf - The buffer to read from.
   * @param prevType - The type of the previous KVP in the same list.
   * @throws ProtocolViolationError if prevType + delta exceeds 2^64 - 1.
   */
  static deserializeDelta(buf: BaseByteBuffer, prevType: bigint): KeyValuePair {
    const deltaType = buf.getVI()
    const typeValue = prevType + deltaType
    if (typeValue > MAX_U64) {
      throw new ProtocolViolationError(
        'KeyValuePair.deserializeDelta',
        `previous type ${prevType} plus delta type ${deltaType} exceeds 2^64 - 1`,
      )
    }
    if (typeValue % 2n === 0n) {
      const value = buf.getVI()
      return new KeyValuePair(typeValue, value)
    } else {
      const len = buf.getNumberVI()
      if (len > MAX_VALUE_LENGTH) {
        throw new LengthExceedsMaxError('KeyValuePair.deserialize', MAX_VALUE_LENGTH, len)
      }
      const value = buf.getBytes(len)
      return new KeyValuePair(typeValue, value)
    }
  }

  /**
   * Checks if this pair is equal to another.
   * @param other - The other KeyValuePair.
   * @returns True if both type and value are equal.
   */
  equals(other: KeyValuePair): boolean {
    if (this.typeValue !== other.typeValue) return false
    if (isVarInt(this) && isVarInt(other)) {
      return this.value === other.value
    }
    if (isBytes(this) && isBytes(other)) {
      const a = this.value
      const b = other.value
      if (a.length !== b.length) return false
      for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) return false
      }
      return true
    }
    return false
  }
}

/**
 * Delta-encodes and serializes a list of Key-Value-Pairs to wire bytes.
 * The list is sorted by ascending type first, since Delta Type is an
 * unsigned varint and cannot represent a type decrease.
 */
export function serializeKvpList(items: KeyValuePair[]): FrozenByteBuffer {
  const sorted = [...items].sort((a, b) => (a.typeValue < b.typeValue ? -1 : a.typeValue > b.typeValue ? 1 : 0))
  const buf = new ByteBuffer()
  let prevType = 0n
  for (const kvp of sorted) {
    buf.putBytes(kvp.serializeDelta(prevType).toUint8Array())
    prevType = kvp.typeValue
  }
  return buf.freeze()
}

/**
 * Reads exactly `count` delta-encoded Key-Value-Pairs from `buf`.
 */
export function deserializeKvpList(buf: BaseByteBuffer, count: number | bigint): KeyValuePair[] {
  const n = typeof count === 'bigint' ? Number(count) : count
  const items: KeyValuePair[] = new Array(n)
  let prevType = 0n
  for (let i = 0; i < n; i++) {
    const kvp = KeyValuePair.deserializeDelta(buf, prevType)
    prevType = kvp.typeValue
    items[i] = kvp
  }
  return items
}

/**
 * Reads delta-encoded Key-Value-Pairs from `buf` until it is exhausted.
 */
export function deserializeKvpListUntilEmpty(buf: BaseByteBuffer): KeyValuePair[] {
  const items: KeyValuePair[] = []
  let prevType = 0n
  while (buf.remaining > 0) {
    const kvp = KeyValuePair.deserializeDelta(buf, prevType)
    prevType = kvp.typeValue
    items.push(kvp)
  }
  return items
}

/**
 * Checks if the KeyValuePair is a varint pair (even typeValue).
 * @param pair - The KeyValuePair to check.
 * @returns True if value is a varint.
 */
export function isVarInt(pair: KeyValuePair): pair is KeyValuePair & { value: bigint } {
  return pair.typeValue % 2n === 0n
}

/**
 * Checks if the KeyValuePair is a blob pair (odd typeValue).
 * @param pair - The KeyValuePair to check.
 * @returns True if value is a Uint8Array.
 */
export function isBytes(pair: KeyValuePair): pair is KeyValuePair & { value: Uint8Array } {
  return pair.typeValue % 2n !== 0n
}

if (import.meta.vitest) {
  const { describe, it, expect } = import.meta.vitest

  describe('KeyValuePair', () => {
    it('roundtrip varint', () => {
      const original = KeyValuePair.tryNewVarInt(2, 100)
      const serialized = original.serialize()
      const parsed = KeyValuePair.deserialize(serialized)
      expect(parsed).toEqual(original)
    })

    it('roundtrip bytes', () => {
      const data = new TextEncoder().encode('test')
      const original = KeyValuePair.tryNewBytes(1, data)
      const serialized = original.serialize()
      const parsed = KeyValuePair.deserialize(serialized)
      expect(parsed).toEqual(original)
    })

    it('invalid type for varint', () => {
      expect(() => KeyValuePair.tryNewVarInt(1, 100)).toThrow(KeyValueFormattingError)
    })

    it('invalid type for bytes', () => {
      const data = new Uint8Array([0x78])
      expect(() => KeyValuePair.tryNewBytes(2, data)).toThrow(KeyValueFormattingError)
    })

    it('length exceeds max', () => {
      const over = new Uint8Array(MAX_VALUE_LENGTH + 1)
      expect(() => KeyValuePair.tryNewBytes(1, over)).toThrow(LengthExceedsMaxError)
    })

    it('deserialize not enough bytes', () => {
      const buf = new ByteBuffer()
      buf.putVI(1) // odd -> bytes variant
      buf.putVI(5) // length = 5
      buf.putBytes(new Uint8Array([0x61, 0x62, 0x63])) // only 3 bytes
      const frozen = buf.freeze()
      expect(() => KeyValuePair.deserialize(frozen)).toThrow(NotEnoughBytesError)
    })

    it('deserialize length casting error', () => {
      const huge = BigInt(Number.MAX_SAFE_INTEGER) + 1n
      const buf = new ByteBuffer()
      buf.putVI(0)
      buf.putVI(huge)
      const frozen = buf.freeze()
      expect(() => KeyValuePair.deserialize(frozen))
    })

    it('deserializeDelta overflow is a protocol violation', () => {
      // previous_type + delta_type must not exceed 2^64 - 1.
      const buf = new ByteBuffer()
      buf.putVI(10n)
      const frozen = buf.freeze()
      expect(() => KeyValuePair.deserializeDelta(frozen, MAX_U64 - 5n)).toThrow(ProtocolViolationError)
    })

    it('serializeDelta with decreasing type is a protocol violation', () => {
      // Delta Type is an unsigned varint, so a type lower than prevType can't be encoded.
      const kvp = KeyValuePair.tryNewVarInt(2, 1)
      expect(() => kvp.serializeDelta(10n)).toThrow(ProtocolViolationError)
    })

    it('serializeKvpList sorts and delta-encodes', () => {
      const items = [
        KeyValuePair.tryNewVarInt(0x20, 1),
        KeyValuePair.tryNewVarInt(0x10, 2),
        KeyValuePair.tryNewBytes(0x21, new TextEncoder().encode('x')),
      ]
      const frozen = serializeKvpList(items)
      const decoded = deserializeKvpList(frozen, 3)
      expect(decoded.map((k) => k.typeValue)).toEqual([0x10n, 0x20n, 0x21n])
    })

    it('deserializeKvpListUntilEmpty reads a delta-encoded stream', () => {
      const items = [KeyValuePair.tryNewVarInt(0x04, 1), KeyValuePair.tryNewVarInt(0x02, 2)]
      const frozen = serializeKvpList(items)
      const decoded = deserializeKvpListUntilEmpty(frozen)
      expect(decoded.map((k) => k.typeValue)).toEqual([0x02n, 0x04n])
    })
  })
}

File: libs/moqtail-rs/src/model/common/pair.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use bytes::{Buf, Bytes, BytesMut};
use std::convert::TryInto;

use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;

const MAX_VALUE_LENGTH: usize = 65535; // 2^16-1

#[derive(Debug, Clone, PartialEq)]
pub enum KeyValuePair {
  VarInt { type_value: u64, value: u64 },
  Bytes { type_value: u64, value: Bytes },
}

impl KeyValuePair {
  /// Fallible constructor for a varint‐typed pair.
  pub fn try_new_varint(type_value: u64, value: u64) -> Result<Self, ParseError> {
    if !type_value.is_multiple_of(2) {
      return Err(ParseError::KeyValueFormattingError {
        context: "KeyValuePair::try_new_varint",
      });
    }
    Ok(KeyValuePair::VarInt { type_value, value })
  }

  /// Fallible constructor for a bytes‐typed pair.
  pub fn try_new_bytes(type_value: u64, value: Bytes) -> Result<Self, ParseError> {
    if type_value.is_multiple_of(2) {
      return Err(ParseError::KeyValueFormattingError {
        context: "KeyValuePair::try_new_bytes",
      });
    }
    let len = value.len();
    if len > MAX_VALUE_LENGTH {
      return Err(ParseError::LengthExceedsMax {
        context: "KeyValuePair::try_new_bytes",
        max: MAX_VALUE_LENGTH,
        len,
      });
    }
    Ok(KeyValuePair::Bytes { type_value, value })
  }

  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    self.serialize_delta(0)
  }

  /// Serializes this pair's wire form, encoding the Type as a delta from
  /// `prev_type` (the type of the previous KVP in the same list, or 0 if
  /// this is the first entry), per draft-ietf-moq-transport-16 1.4.2.
  pub fn serialize_delta(&self, prev_type: u64) -> Result<Bytes, ParseError> {
    let delta_type = self.get_type().checked_sub(prev_type).ok_or_else(|| {
      ParseError::ProtocolViolation {
        context: "KeyValuePair::serialize_delta",
        details: format!(
          "type {} is less than previous type {prev_type}; a KVP list must be sorted by ascending type to be delta-encoded",
          self.get_type()
        ),
      }
    })?;

    let mut buf = BytesMut::new();
    buf.put_vi(delta_type)?;
    match self {
      Self::VarInt { value, .. } => {
        buf.put_vi(*value)?;
      }
      Self::Bytes { value, .. } => {
        buf.put_vi(value.len() as u64)?;
        buf.extend_from_slice(value);
      }
    }
    Ok(buf.freeze())
  }

  pub fn deserialize(bytes: &mut Bytes) -> Result<Self, ParseError> {
    Self::deserialize_delta(bytes, 0)
  }

  /// Deserializes a wire-form KVP whose Type field is a delta from
  /// `prev_type` (the type of the previous KVP in the same list, or 0 if
  /// this is the first entry), per draft-ietf-moq-transport-16 1.4.2.
  pub fn deserialize_delta(bytes: &mut Bytes, prev_type: u64) -> Result<Self, ParseError> {
    let delta_type = bytes.get_vi()?;
    let type_value =
      prev_type
        .checked_add(delta_type)
        .ok_or_else(|| ParseError::ProtocolViolation {
          context: "KeyValuePair::deserialize_delta",
          details: format!(
            "previous type {prev_type} plus delta type {delta_type} exceeds 2^64 - 1"
          ),
        })?;

    if type_value % 2 == 0 {
      // VarInt variant
      let value = bytes.get_vi()?;
      Ok(KeyValuePair::VarInt { type_value, value })
    } else {
      // Bytes variant
      let len_u64 = bytes.get_vi()?;
      let len: usize =
        len_u64
          .try_into()
          .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
            context: "KeyValuePair::deserialize length",
            from_type: "u64",
            to_type: "usize",
            details: e.to_string(),
          })?;

      if len > MAX_VALUE_LENGTH {
        return Err(ParseError::LengthExceedsMax {
          context: "KeyValuePair::deserialize",
          max: MAX_VALUE_LENGTH,
          len,
        });
      }
      if bytes.remaining() < len {
        return Err(ParseError::NotEnoughBytes {
          context: "KeyValuePair::deserialize value",
          needed: len,
          available: bytes.remaining(),
        });
      }

      let value = bytes.copy_to_bytes(len);

      Ok(KeyValuePair::Bytes { type_value, value })
    }
  }

  pub fn is_same_type(&self, other: &KeyValuePair) -> bool {
    self.get_type() == other.get_type()
  }

  pub fn get_type(&self) -> u64 {
    match self {
      KeyValuePair::VarInt { type_value, .. } => *type_value,
      KeyValuePair::Bytes { type_value, .. } => *type_value,
    }
  }
}

/// Delta-encodes and serializes a list of Key-Value-Pairs to wire bytes.
/// The list is sorted by ascending type first, since Delta Type is an
/// unsigned varint and cannot represent a type decrease.
pub fn serialize_kvp_list(items: &[KeyValuePair]) -> Result<Bytes, ParseError> {
  let mut sorted: Vec<&KeyValuePair> = items.iter().collect();
  sorted.sort_by_key(|kvp| kvp.get_type());

  let mut buf = BytesMut::new();
  let mut prev_type = 0u64;
  for kvp in sorted {
    buf.extend_from_slice(&kvp.serialize_delta(prev_type)?);
    prev_type = kvp.get_type();
  }
  Ok(buf.freeze())
}

/// Reads exactly `count` delta-encoded Key-Value-Pairs from `bytes`.
pub fn deserialize_kvp_list(
  bytes: &mut Bytes,
  count: u64,
) -> Result<Vec<KeyValuePair>, ParseError> {
  let mut items = Vec::with_capacity(count as usize);
  let mut prev_type = 0u64;
  for _ in 0..count {
    let kvp = KeyValuePair::deserialize_delta(bytes, prev_type)?;
    prev_type = kvp.get_type();
    items.push(kvp);
  }
  Ok(items)
}

/// Reads delta-encoded Key-Value-Pairs from `bytes` until it is exhausted.
pub fn deserialize_kvp_list_until_empty(
  bytes: &mut Bytes,
) -> Result<Vec<KeyValuePair>, ParseError> {
  let mut items = Vec::new();
  let mut prev_type = 0u64;
  while bytes.has_remaining() {
    let kvp = KeyValuePair::deserialize_delta(bytes, prev_type)?;
    prev_type = kvp.get_type();
    items.push(kvp);
  }
  Ok(items)
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::{Bytes, BytesMut};

  #[test]

  fn roundtrip_varint() {
    let original = KeyValuePair::try_new_varint(2, 100).unwrap();
    let mut buf = original.serialize().unwrap();
    let parsed = KeyValuePair::deserialize(&mut buf).unwrap();
    assert_eq!(parsed, original);
  }

  #[test]
  fn roundtrip_bytes() {
    let original = KeyValuePair::try_new_bytes(1, Bytes::from("test")).unwrap();
    let mut buf = original.serialize().unwrap();
    let parsed = KeyValuePair::deserialize(&mut buf).unwrap();
    assert_eq!(parsed, original);
  }

  #[test]
  fn invalid_type_varint() {
    let err = KeyValuePair::try_new_varint(1, 100);
    assert!(err.is_err());
  }

  #[test]
  fn invalid_type_bytes() {
    let err = KeyValuePair::try_new_bytes(2, Bytes::from("x"));
    assert!(err.is_err());
  }

  #[test]
  fn length_exceeds_max() {
    let data = Bytes::from(vec![0u8; MAX_VALUE_LENGTH + 1]);
    let err = KeyValuePair::try_new_bytes(1, data).unwrap_err();
    assert!(matches!(err, ParseError::LengthExceedsMax { .. }));
  }

  #[test]
  fn deserialize_not_enough_bytes() {
    let mut buf = BytesMut::new();
    buf.put_vi(1).unwrap(); // odd -> bytes variant
    buf.put_vi(5).unwrap(); // length = 5
    buf.extend_from_slice(b"abc"); // only 3 bytes
    let mut bytes = buf.freeze();
    let err = KeyValuePair::deserialize(&mut bytes).unwrap_err();
    assert!(matches!(err, ParseError::NotEnoughBytes { .. }));
  }

  #[test]
  fn is_same_type_test() {
    let kv1 = KeyValuePair::try_new_varint(2, 100).unwrap();
    let kv2 = KeyValuePair::try_new_varint(2, 200).unwrap();
    let kv3 = KeyValuePair::try_new_bytes(1, Bytes::from("test")).unwrap();
    assert!(kv1.is_same_type(&kv2));
    assert!(!kv1.is_same_type(&kv3));
  }

  #[test]
  fn get_type_test() {
    let kv1 = KeyValuePair::try_new_varint(2, 100).unwrap();
    let kv2 = KeyValuePair::try_new_bytes(1, Bytes::from("test")).unwrap();
    assert_eq!(kv1.get_type(), 2);
    assert_eq!(kv2.get_type(), 1);
  }

  #[test]
  fn deserialize_delta_overflow_is_protocol_violation() {
    // previous_type + delta_type must not exceed u64::MAX per draft-16 1.4.2.
    let mut buf = BytesMut::new();
    buf.put_vi(10u64).unwrap();
    let mut bytes = buf.freeze();
    let err = KeyValuePair::deserialize_delta(&mut bytes, u64::MAX - 5).unwrap_err();
    assert!(matches!(err, ParseError::ProtocolViolation { .. }));
  }

  #[test]
  fn serialize_delta_decreasing_type_is_protocol_violation() {
    // Delta Type is an unsigned varint, so a type lower than prev_type can't be encoded.
    let kvp = KeyValuePair::try_new_varint(2, 1).unwrap();
    let err = kvp.serialize_delta(10).unwrap_err();
    assert!(matches!(err, ParseError::ProtocolViolation { .. }));
  }

  #[test]
  fn serialize_kvp_list_sorts_and_delta_encodes() {
    let items = vec![
      KeyValuePair::try_new_varint(0x20, 1).unwrap(),
      KeyValuePair::try_new_varint(0x10, 2).unwrap(),
      KeyValuePair::try_new_bytes(0x21, Bytes::from_static(b"x")).unwrap(),
    ];
    let mut bytes = serialize_kvp_list(&items).unwrap();
    let decoded = deserialize_kvp_list(&mut bytes, 3).unwrap();
    let types: Vec<u64> = decoded.iter().map(|k| k.get_type()).collect();
    assert_eq!(types, vec![0x10, 0x20, 0x21]);
  }

  #[test]
  fn deserialize_kvp_list_until_empty_reads_delta_encoded_stream() {
    let items = vec![
      KeyValuePair::try_new_varint(0x04, 1).unwrap(),
      KeyValuePair::try_new_varint(0x02, 2).unwrap(),
    ];
    let mut bytes = serialize_kvp_list(&items).unwrap();
    let decoded = deserialize_kvp_list_until_empty(&mut bytes).unwrap();
    let types: Vec<u64> = decoded.iter().map(|k| k.get_type()).collect();
    assert_eq!(types, vec![0x02, 0x04]);
  }
}
